jaxrs-spec/src/main/asciidoc/chapters/providers/_lifecycle_and_environment.adoc

```diff
@@ -33,6 +33,13 @@ discover provider classes via mechanisms such as class scanning. A
 JAX-RS implementation that supports automatic discovery of classes MUST
 process only those classes that are annotated with `@Provider`.
 
+[[services]]
+==== Services
+An implementation MUST automatically register the following set of `jakarta.ws.rs.core.Feature` providers in `jakarta.ws.rs.core.Configuration` runtime configurations:
+
+* if the JAX-RS runtime was loaded in a `java.lang.ModuleLayer` L, all providers discovered by `java.util.ServiceLoader.load(L, Feature.class)`
+* else, all providers discovered by `ServiceLoader.load​(Feature.class, Feature.class.getClassLoader())`
```

Comment:
I think we would not want to use the Feature's classloader.  Instead, we should follow the pattern we use in the `FactoryFinder` classes where we first try the thread context classloader, then `Feature.class.getClassLoader()`, and then possibly a `jaxrs.properties` file. 

The trouble with just using the `Feature` class is that multi-app containers like Liberty and Wildfly might allow users to add their extensions/services/whatever-we-call-it to their app's classpath, but then it would not be found by the Feature's classloader.

Reply:
That's a good point by @andymc12.

Reply:
The rationale for just using `Feature`'s `ClassLoader` is to prevent undesired auto-loading of `Feature`s in EARs/WARs.

As I understood it, this feature is about extending the JAX-RS runtime in a portable way & without involvement of the application developer. So I'd argue `Feature`s must not be auto-loaded from developer-owned artifacts such as EARs/WARs.

Currently, developers have total control over what is and isn't loaded from their artifacts via `Application`. And if desired, they can use `ServiceLoader` to find & register additional `Feature`s themselves using the thread's context `ClassLoader`.

By requiring auto-loading via the thread's context `ClassLoader`, developers would loose a certain degree of control and it might lead to subtle, hard-to-debug issues (see also [this comment](https://github.com/eclipse-ee4j/jaxrs-api/pull/933#issuecomment-755449932) e.g.). In this case, developers would need a way to opt-out of this behavior, which would complicate the feature.

What do you think?

Reply:
>As I understood it, this feature is about extending the JAX-RS runtime in a portable way & without involvement of the application developer. So I'd argue Features must not be auto-loaded from developer-owned artifacts such as EARs/WARs.

Yes and no. In an embedded application scenario (like using Quarkus or a Java SE environment), everything would be on the same classpath so the JAX-RS implementation code wouldn't know whether a particular extension feature came from the application developer or as a product extension - it's just on the classpath.  But then if a developer did package their extension with the app and then ran it in multi-app container (like Wildfly/Liberty/etc.) they would see different results - so in effect, this approach isn't portable.

> Currently, developers have total control over what is and isn't loaded from their artifacts via Application. And if desired, they can use ServiceLoader to find & register additional Features themselves using the thread's context ClassLoader.

True, but that is not intuitive.  I don't think I've ever seen an example where a user used the ServiceLoader in the Application subclass.  Further, we are deprecating and (eventually) removing the `getSingletons` method from `Application`, which will make it that much harder for users to add providers via the ServiceLoader pattern - they would need to use `getClasses` which means using the ServiceLoader to return an instance of the provider, and the return the instance's class.  

> By requiring auto-loading via the thread's context ClassLoader, developers would loose a certain degree of control and it might lead to subtle, hard-to-debug issues (see also this comment e.g.). In this case, developers would need a way to opt-out of this behavior, which would complicate the feature.

I think this is a problem with the idea of extensions in general - regardless of which classloader is used. For example, a developer will run and (hopefully) pass tests on their application before turning it over to production. But if the production admins decides to add some new monitoring extension to the container, it might change the behavior of the app in unexpected ways.  I would think that if the extension were part of the application, it would actually make it easier to find and fix the behavior differences earlier in the development lifecycle.

But even aside from embedding the extensions in the application archives, these multi-app servers also allow apps to use shared libraries which is the deployment target of these extensions - and the shared libraries are accessed via the thread context classloader.

Reply:
@anthonyvdotbe Are you referring to the situation where there's a jar in (say) an application server with a `META-INF/jakarta.ws.rs.core.Feature` that lists "foo.bar.Kaz", that there might be an EAR or WAR out there that contains a class that's by coincidence also called "foo.bar.Kaz", and that the REST runtime then loads the "foo.bar.Kaz" from the application archive instead of the application server archive?

Reply:
@arjantijms I'm referring to the concerns raised by @jansupol [here](https://github.com/eclipse-ee4j/jaxrs-api/pull/933#issuecomment-755449932) w.r.t. an EAR containing multiple modules developed by independent teams, where the module of one team could interfere with the modules of other teams. Or even just: a WAR, where one of the transitive dependencies could at any point add a `Feature` that has undesired (or even malicious) effects, even when an `Application` subclass is present.

Reply:
I thought the main use case for these extensions is precisely that in which the features are _not_ part of an application (much like CDI extensions). If it is part of an application, it can be added without any auto-loading mechanism. The corresponding facility in Jersey is mostly used by frameworks that sit on top of Jakarta REST. Thus, this needs to follow the `FactoryFinder` approach as pointed out by @andymc12 to make sure it runs predictably in all environments.

I also think the proposal should handle `DynamicFeature` or explain why it is left out at least.